core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+pub const POW_AUX_PREFIX: [u8; 4] = *b"PoW:";
+
+fn aux_key(hash: &H256) -> Vec<u8> {
+	POW_AUX_PREFIX.iter().chain(&hash[..])
+		.cloned().collect::<Vec<_>>()
+}
+
+/// Auxiliary data for PoW.
+#[derive(Encode, Decode, Clone, Debug, Default)]
+pub struct PowAux {
+	/// Total difficulty.
+	pub total_difficulty: Difficulty,
+}
+
+impl PowAux {
+	/// Read the auxiliary from client.
+	pub fn read<C: AuxStore>(client: &C, hash: &H256) -> Result<Self, String> {
+		let key = aux_key(hash);
+
+		match client.get_aux(&key).map_err(|e| format!("{:?}", e))? {
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e)),
+			None => Ok(Default::default()),
+		}
+	}
+}
+
+/// Algorithm used for proof of work.
+pub trait PowAlgorithm<B: BlockT> {
+	/// Get the next block's difficulty.
+	fn difficulty(&self, parent: &BlockId<B>) -> Result<Difficulty, String>;
+	/// Verify proof of work against the given difficulty.
+	fn verify(
+		&self,
+		parent: &BlockId<B>,
+		pre_hash: &H256,
+		seal: &Seal,
+		difficulty: Difficulty
+	) -> Result<bool, String>;
+	/// Mine a seal that satisfy the given difficulty.
+	fn mine(
+		&self,
+		parent: &BlockId<B>,
+		pre_hash: &H256,
+		seed: &H256,
+		difficulty: Difficulty,
+		round: u32
+	) -> Result<Option<Seal>, String>;
+}
+
+/// Delegates algorithm selection to Substrate runtime.
+pub struct PowRuntimeAlgorithm<C> {
+	client: Arc<C>,
+}
+
+impl<C> PowRuntimeAlgorithm<C> {
+	/// Create a new runtime algorithm given a client.
+	pub fn new(client: Arc<C>) -> Self {
+		Self { client }
+	}
+}
+
+impl<B: BlockT<Hash=H256>, C> PowAlgorithm<B> for PowRuntimeAlgorithm<C> where
+	C: ProvideRuntimeApi,
+	C::Api: PowApi<B>,
+{
+	fn difficulty(&self, parent: &BlockId<B>) -> Result<Difficulty, String> {
+		self.client.runtime_api().difficulty(parent)
+			.map_err(|e| format!("Fetch difficulty failed: {:?}", e))
+	}
+
+	fn verify(
+		&self,
+		parent: &BlockId<B>,
+		pre_hash: &H256,
+		seal: &Seal,
+		difficulty: Difficulty
+	) -> Result<bool, String> {
+		self.client.runtime_api().verify(parent, pre_hash, seal, difficulty)
+			.map_err(|e| format!("Verify PoW seal failed: {:?}", e))
+	}
+
+	fn mine(
+		&self,
+		parent: &BlockId<B>,
+		pre_hash: &H256,
+		seed: &H256,
+		difficulty: Difficulty,
+		round: u32
+	) -> Result<Option<Seal>, String> {
+		self.client.runtime_api().mine(parent, pre_hash, seed, difficulty, round)
+			.map_err(|e| format!("Mine PoW seal failed: {:?}", e))
+	}
+}
+
+/// A verifier for PoW blocks.
+pub struct PowVerifier<C, Algorithm> {
+	client: Arc<C>,
+	algorithm: Algorithm,
+	inherent_data_providers: inherents::InherentDataProviders,
+}
+
+impl<C, Algorithm> PowVerifier<C, Algorithm> {
+	fn check_header<B: BlockT<Hash=H256>>(
+		&self,
+		mut header: B::Header,
+		block_id: BlockId<B>,
+	) -> Result<(B::Header, Difficulty, DigestItem<H256>), String> where
+		Algorithm: PowAlgorithm<B>,
+	{
+		let hash = header.hash();
+
+		let (seal, inner_seal) = match header.digest_mut().pop() {
```

Comment:
Dumb question: what guarantees that the pow seal is the last item?

Reply:
I think we actually guarantee that it's the last item for pow engine. We only pushes seal after all the rest of the block finished building. It also must be the last block (and if not we should reject it as invalid), because otherwise, the prehash we got will be incorrect.

Reply:
@sorpass That is correct.

Reply:
I think this info should be in a comment above the `let` binding.